yarp/yarp_compiler.c

```diff
@@ -1503,6 +1503,60 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
           ADD_SETLOCAL(ret, &dummy_line_node, (int)index, local_write_node->depth);
           return;
       }
+      case YP_MATCH_WRITE_NODE: {
+          yp_match_write_node_t *cast = (yp_match_write_node_t *)node;
+          LABEL *fail_label = NEW_LABEL(lineno);
+          LABEL *end_label = NEW_LABEL(lineno);
+          size_t capture_count = cast->locals.size;
+
+          YP_COMPILE((yp_node_t *)cast->call);
+          VALUE global_variable_name = rb_id2sym(rb_intern("$~"));
+
+          ADD_INSN1(ret, &dummy_line_node, getglobal, global_variable_name);
+          ADD_INSN(ret, &dummy_line_node, dup);
+          ADD_INSNL(ret, &dummy_line_node, branchunless, fail_label);
+
+          for (size_t index = 0; index < capture_count; index++) {
+              yp_constant_id_t constant = cast->locals.ids[index];  
+              int local_index = yp_lookup_local_index(iseq, compile_context, constant);
+
+              if (index < (capture_count - 1)) {
+                  ADD_INSN(ret, &dummy_line_node, dup);
+              }
+              ADD_INSN1(ret, &dummy_line_node, putobject, rb_id2sym(yp_constant_id_lookup(compile_context, constant)));
+              ADD_SEND(ret, &dummy_line_node, rb_intern("[]"), INT2FIX(1));
+
+              if (capture_count == 1) {
+                  DECL_ANCHOR(nom);
+                  INIT_ANCHOR(nom);
+
+                  ADD_INSNL(nom, &dummy_line_node, jump, end_label);
+                  ADD_LABEL(nom, fail_label);
+                  ADD_LABEL(nom, end_label);
+                  ADD_SETLOCAL(nom, &dummy_line_node, local_index, 0);
+
+                  ADD_SEQ(ret, nom);
+                  return;
+              } else {
+                  ADD_SETLOCAL(ret, &dummy_line_node, local_index, 0);
+              }
+                  
+          }
+
+          ADD_INSNL(ret, &dummy_line_node, jump, end_label);
+          ADD_LABEL(ret, fail_label);
+          ADD_INSN(ret, &dummy_line_node, pop);
+
+          for (size_t index = 0; index < capture_count; index++) {
+              yp_constant_id_t constant = cast->locals.ids[index];  
+              int local_index = yp_lookup_local_index(iseq, compile_context, constant);
```

Comment:
It looks like we're fetching the index for all of the constants twice (once in each loop). Instead, could we stack allocate a list that mirrors the constants? You should be able to use `ALLOCV_N` for this.